### PR TITLE
8024 mdb_ctf_vread() needn't be so strict about unions

### PR DESCRIPTION
**Pull request by prakashsurya**
Reviewed by: Matthew Ahrens <mahrens@delphix.com>
Reviewed by: Pavel Zakharov <pavel.zakharov@delphix.com>

This modifies mdb's vread_helper function such that not all target union
members need to be present in the mdb type. If there is only a single
union member in the mdb type, its actual type does not need to match
with its target's type. On the other hand, if more than one union
members are specified in the mdb type, their types must match with the
types of their relevant union members of the target union.

Upstream bugs: DLPX-23010

### CODE REVIEW

**Comment by prakashsurya** [comment]
@zettabot go

File: usr/src/cmd/mdb/common/mdb/mdb_ctf.c [hunk 1]
```diff
@@ -1468,29 +1468,52 @@ vread_helper(mdb_ctf_id_t modid, char *modbuf,
...
-		 * Unions are a little tricky. The only time it's truly
-		 * safe to read in a union is if no part of the union or
-		 * any of its component types have changed.  The correct
-		 * use of this feature is to read the containing structure,
-		 * figure out which component of the union is valid, compute
-		 * the location of that in the target and then read in
-		 * that part of the structure.
+		 * Not all target union members need to be present in the
+		 * mdb type. If there is only a single union member in the
+		 * mdb type, its actual type does not need to match with
```
Comment (rmustacc) [inline usr/src/cmd/mdb/common/mdb/mdb_ctf.c]: It'd really help if it explained _why_ this is OK. Can someone elaborate on what makes this safe?

Comment (rmustacc) [inline usr/src/cmd/mdb/common/mdb/mdb_ctf.c]: OK, I think I see what you're saying. Basically if the target member is there then the encapsulating union and its sizing doesn't matter, as we're just reading it into the target type, that right?

Comment (ahrens) [inline usr/src/cmd/mdb/common/mdb/mdb_ctf.c]: Right.

Comment (ahrens) [inline usr/src/cmd/mdb/common/mdb/mdb_ctf.c]: @rmustacc Are you OK with this explanation and can we count you as a reviewer?

Comment (rmustacc) [inline usr/src/cmd/mdb/common/mdb/mdb_ctf.c]: Sorry, I missed the reply. Yeah, I'm fine. Thanks Matt!

**Pull request closed without merging**